cmd/training-operator.v1/main.go

```diff
@@ -182,15 +198,52 @@ func setupControllers(mgr ctrl.Manager, enabledSchemes controllerv1.EnabledSchem
 	}
 	errMsg := "failed to set up controllers"
 	for _, s := range enabledSchemes {
-		setupFunc, supported := controllerv1.SupportedSchemeReconciler[s]
-		if !supported {
+		setupReconcilerFunc, supportedReconciler := controllerv1.SupportedSchemeReconciler[s]
+		if !supportedReconciler {
 			setupLog.Error(errors.New(errMsg), "scheme is not supported", "scheme", s)
 			os.Exit(1)
 		}
-		if err := setupFunc(mgr, gangSchedulingSetupFunc, controllerThreads); err != nil {
+		if err := setupReconcilerFunc(mgr, gangSchedulingSetupFunc, controllerThreads); err != nil {
 			setupLog.Error(errors.New(errMsg), "unable to create controller", "scheme", s)
 			os.Exit(1)
 		}
+		setupWebhookFunc, supportedWebhook := webhooks.SupportedSchemeWebhook[s]
+		if !supportedWebhook {
+			setupLog.Error(errors.New(errMsg), "scheme is not supported", "scheme", s)
+			os.Exit(1)
+		}
+		if err := setupWebhookFunc(mgr); err != nil {
+			setupLog.Error(errors.New(errMsg), "unable to start webhook server", "scheme", s)
+			os.Exit(1)
+		}
+	}
+}
+
+func setupProbeEndpoints(mgr ctrl.Manager, certsReady <-chan struct{}) {
+	defer setupLog.Info("Probe endpoints are configured on healthz and readyz")
+
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+
+	// Wait for the webhook server to be listening before advertising the
+	// training-operator replica as ready. This allows users to wait with sending the first
+	// requests, requiring webhooks, until the training-operator deployment is available, so
+	// that the early requests are not rejected during the traininig-operator's startup.
+	// We wrap the call to GetWebhookServer in a closure to delay calling
+	// the function, otherwise a not fully-initialized webhook server (without
+	// ready certs) fails the start of the manager.
+	if err := mgr.AddReadyzCheck("readyz", func(req *http.Request) error {
+		select {
+		case <-certsReady:
+			return mgr.GetWebhookServer().StartedChecker()(req)
```

Comment:
Can you explain about this closure?  how is  `StartedChecker()` working with certsReady logic? `StartedChecked` checks if webhook server is up while `certsReady` is called here https://github.com/open-policy-agent/cert-controller/blob/master/pkg/rotator/rotator.go#L888

Reply:
@johnugeorge If we don't have this extended readiness probe, the controller-runtime readiness server responds "Ready" even if the training operator is still preparing due to certificate generation.
This extended check allows us to check if the training-operator is ready correctly.

Also, you can find the details here: https://github.com/kubernetes-sigs/kueue/pull/1676/files#r1472752662